apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.YCSB/Program.cs

```diff
@@ -0,0 +1,513 @@
+                .WaitWithUnwrappedExceptions();
+
+            DebugMessage("running...");
+            var overallTime = Stopwatch.StartNew();
+            if (s_options.ContainsKey(s_targetQps))
+            {
+                //target qps version
+                RunWorkerWithQps(GetOption<double>(s_targetQps), TimeSpan.FromSeconds(GetOption<int>(s_qpstestTime)))
+                    .WaitWithUnwrappedExceptions();
+            }
+            else
+            {
+                //target worker version
+                RunWithTargetWorkers(GetOption<int>(s_num_worker)).WaitWithUnwrappedExceptions();
+            }
+            PrintResults(overallTime.ElapsedMilliseconds);
+
+            return 0;
+        }
+
+        /// <summary>
+        /// Runs a single worker task
+        /// </summary>
+        private static async Task PreWarm()
+        {
+            using (var spannerConnection = new SpannerConnection(
+                $"Data Source=projects/{s_project}/instances/{s_instance}/databases/{s_database}"))
+            {
+                await spannerConnection.OpenAsync();
+            }
+        }
+
+        /// <summary>
+        /// Runs a single worker task that targets a specific QPS without regard to workers.
+        /// (this is from C# spanner stress tests)
+        /// </summary>
+        private static async Task RunWorkerWithQps(double queriesPerSecond, TimeSpan testTime)
+        {
+            var sw = Stopwatch.StartNew();
+            var all = new List<Task>();
+
+            while (sw.Elapsed < testTime)
+            {
+                if (sw.Elapsed.TotalSeconds * queriesPerSecond > all.Count)
+                {
+                    var operationTarget = s_rand.Value.NextDouble() * s_total;
+                    var operation = 0;
+                    operationTarget -= s_operations[operation].Proportion;
+                    while (operationTarget > 0 && operation < s_operations.Count)
+                    {
+                        operation++;
+                        if (operation < s_operations.Count)
+                        {
+                            operationTarget -= s_operations[operation].Proportion;
+                        }
+                    }
+                    try
+                    {
+                        all.Add(s_operations[Math.Min(operation, s_operations.Count - 1)].Action(
+                            s_operations[Math.Min(operation, s_operations.Count - 1)].Latencies));
+                    }
+                    catch (Exception e)
+                    {
+                        DebugMessage($"This error is being ignored and we are continuing the test: {e.Message}");
+                    }
+                }
+            }
+
+            var timeout = Task.Delay(TimeSpan.FromSeconds(10));
+            await Task.WhenAny(Task.WhenAll(all), timeout);
+            s_totaloperations = all.Count;
+        }
+
+        /// <summary>
+        /// Runs a single worker task
```

Comment:
Doesn't it run `numWorkers` tasks?

Reply:
I've clarified the comment.  It's implementation changed slightly to reduce some awaits.